File: Write_test/PDD/2/1.py
# nums = list(map(int, input().split()))
# cars = sorted(nums, key=lambda x: x[0], reverse=1)


n = int(input())
size = n//2
result = [[0 for j in range(n)]for i in range(n)]

mapping = {
    (1, -1, 1, 1): 1,
    (1, 1, 1, 1): 2,
    (1, 1, 1, -1): 3,
    (-1, 1, 1, -1): 4,
    (-1, 1, -1, -1): 5,
    (-1, -1, -1, -1): 6,
    (-1, -1, -1, 1): 7,
    (1, -1, -1, 1): 8
}


def up(i, j):
    if i < size:
        return 1
    elif i>=n-size:
        return -1
    else:
        return 0


def left(i, j):
    if j < size:
        return 1
    elif j>=n-size:
        return -1
    else:
        return 0


def lu(i, j):
    if i+j < n-1:
        return 1
    elif i+j>n-1:
        return -1
    else:
        return 0


def ru(i, j):
    if j > i:
        return 1
    elif j<i:
        return -1
    else:
        return 0


for i in range(n):
    for j in range(n):
        tempkey = (up(i, j), left(i, j), lu(i, j), ru(i, j))
        if tempkey in mapping.keys():
            result[i][j] = mapping[tempkey]

for nums in result:
    tempstr=" ".join(map(str,nums))
    print(tempstr)

File: Write_test/wangyi/huyu/1.cpp
// # nums = list(map(int, input().split()))
// # cars = sorted(nums, key=lambda x: x[0], reverse=True)

// # " ".join(map(str,nums))
// boxNum, personNum = list(map(int, input().split()))
// boxNum += 1
// boxValueList = list(map(int, input().split()))
// boxValueList.insert(0, 0)
// boxStatusList = [[]for i in range(boxNum)]
// for per in range(personNum):
//     actionNum = int(input())
//     personnalPrice = 0
//     leftHand = 0
//     rightHand = 0
//     bag = []
//     for actIndex in range(actionNum):
//         act = list(input().split())
//         if act[1] == 'keep':
//             if act[0] == 'left':
//                 bag.append(str(leftHand))
//                 leftHand = 0
//             else:
//                 bag.append(str(rightHand))
//                 rightHand = 0
//         elif act[1] == 'take':
//             target = int(act[2])
//             takeResult = target
//             if len(boxStatusList[target]):
//                 takeResult = boxStatusList[target].pop(-1)
//             if act[0] == 'left':
//                 leftHand = takeResult
//             else:
//                 rightHand = takeResult
//         else:
//             if act[0] == 'left':
//                 boxStatusList[int(act[2])].append(leftHand)
//                 leftHand = 0
//             else:
//                 boxStatusList[int(act[2])].append(rightHand)
//                 rightHand = 0
//     for item in bag:
//         personnalPrice += boxValueList[int(item)]
//     print(personnalPrice)
#include <iostream>
#include <vector>
#include <tuple>
#include <algorithm>
#include <unordered_map>
#include <array>
#include <numeric>
#include <stack>
#include <string>
using namespace std;


int main() {
    int boxNum, personNum;
    cin >> boxNum >> personNum;
    vector<int> boxValueList{ 0 };
    vector<stack<int>> boxStatus{ stack<int>{} };

    for (int i = 0; i < boxNum; i++) {
        int tmp;
        cin >> tmp;
        boxValueList.push_back(tmp);
        boxStatus.push_back(stack<int>{});
    }
    for (int i = 0; i < personNum; i++) {
        vector<int> bag{};
        int leftHand{}, rightHand{};
        int actNum{};
        cin >> actNum;
        for (int j = 0; j < actNum; j++) {
            string hand, act, target;
            cin >> hand >> act;
            if (act != "keep")
                cin >> target;
            if (act == "keep") {
                if (hand == "left") {
                    //atoi(target.c_str())
                    bag.push_back(leftHand);
                    leftHand = 0;
                }
                else {
                    bag.push_back(rightHand);
                    rightHand = 0;
                }
            }
            else if (act == "take") {
                int takeResult = stoi(target);
                if (!boxStatus[stoi(target)].empty()) {
                    takeResult = boxStatus[stoi(target)].top();
                    boxStatus[stoi(target)].pop();
                }
                if (hand == "left")
                    leftHand = takeResult;
                else
                    rightHand = takeResult;
            }
            else {
                if (hand == "left") {
                    boxStatus[stoi(target)].push(leftHand);
                    leftHand = 0;
                }
                else {
                    boxStatus[stoi(target)].push(rightHand);
                    rightHand = 0;
                }
            }
        }
        int result{};
        for (auto choose : bag) {
            result += boxValueList[choose];
        }
        if (leftHand!=0)
            result+= boxValueList[leftHand];
        if (rightHand!=0)
            result += boxValueList[rightHand];
        cout << result<<endl;
    }
    return 0;


}